### PR TITLE
Registration Frontend in React with Authentication and Validation

### PR DESCRIPTION
**Pull request by musanaeem**
Using Django-Rest backend APIs, I created a Registration Frontend with Field validations onBlur. The submit button on the register form will remain disabled till all fields are filled and validated. Validation for Date of Birth does not exist in the current PR. 

- registration form returns appropriate errors, not only for validation but also when a request of an existing user is sent and when registration succeeds and when it fails
- Form validation was converted to React code with the use of useState hooks.
- Upon successful registration, the user is redirected to login page.
- The styling is the same as the initial Django project

Screenshots:

Validation Errors:
image

User Already Exists:
image

Successful Registration:
image

### CODE REVIEW

File: src/components/FormContainer.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+                <div className="mt-4">
+                    <div className="d-flex justify-content-center links">
+                        {accountMessage} <Link to={link} className="ml-2">{alternateTitle}</Link>
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+  )
+}
```
Comment (srehaider) [inline src/components/FormContainer.js]: Use all traditional functions are all arrow functions.

Use children prop to render the form component.
```suggestion
function FormContainer({link, title, form, accountMessage, alternateTitle, children}) {
  return (

    <div className="container h-100">
        <div className="d-flex justify-content-center h-100">
            <div className="user-card">
                <div className="d-flex justify-content-center">
                    <h3 className="form-title">{title}</h3>
                </div>

                <div className="d-flex justify-content-center form-container">
                    {children}
                </div>

                <div className="mt-4">
                    <div className="d-flex justify-content-center links">
                        {accountMessage} <Link to={link} className="ml-2">{alternateTitle}</Link>
                    </div>
                </div>
            </div>
        </div>
    </div>
  )
}
```

File: src/components/RegisterForm.js [hunk 2]
```diff
@@ -0,0 +1,271 @@
+import React, { useState } from 'react'
```
Comment (srehaider) [inline src/components/RegisterForm.js:1]: I still need to review this file!

File: src/pages/HomePage.js [hunk 3]
```diff
@@ -1,10 +1,16 @@
 import React, { Component } from 'react'
 import { useNavigate } from 'react-router-dom';
+import logoutRequest from '../services/LogoutRequest';
 
 class HomePage extends Component {
+
+  constructor(props){
+  super(props);
+}
```
Comment (srehaider) [inline src/pages/HomePage.js]: Fix indentation.

File: src/pages/HomePage.js [hunk 4]
```diff
@@ -1,10 +1,16 @@
...
   
   logoutUser = () => {
     window.localStorage.setItem('blogSiteUserLoggedIn', false);
+    logoutRequest().then();
     this.props.changeLoginState(false);
     this.props.navigator('/login');
```
Comment (srehaider) [inline src/pages/HomePage.js]:
```suggestion
    logoutRequest().then(() => {
        this.props.changeLoginState(false);
        this.props.navigator('/login');
    });
```

File: src/pages/RegisterPage.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            }
+        });
+    }
+    
+
+
+    return (
+        <div>
+            <FormContainer form={registerTag} title='Register' alternateTitle='Login'  accountMessage='Already have an account?' link='/login'/>
+        </div>
```
Comment (srehaider) [inline src/pages/RegisterPage.js]: Use children prop
```suggestion
        <FormContainer title='Register' alternateTitle='Login'  accountMessage='Already have an account?' link='/login'>
            <RegisterForm onRegister = {registerUser} error = {error}/>
        </FormContainer>
```

File: src/services/LogoutRequest.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        'Content-Type': 'application/json',
+        'Authorization': null,
+        'X-CSRFToken': Cookies.get('csrftoken')
+      },
+      credentials: 'include',
+      body: "",
+    };
+
+    return options
+  }
```
Comment (srehaider) [inline src/services/LogoutRequest.js]: This code is common among all services only the body is different, you can extract this code as a helper/utility function and use it in all services.

File: src/services/LogoutRequest.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+function LogoutRequest() {
+
+    let options = getOptions()
+
+        const env = process.env;
+
+        const url = `${env.REACT_APP_HOST}:${env.REACT_APP_PORT}${env.REACT_APP_LOGOUT_PATH}`
```
Comment (srehaider) [inline src/services/LogoutRequest.js]: Better to add the port in `REACT_APP_HOST`, on production we don't have different ports, the default ports 80 or 443 are used and we don't mention them in the URL.

**Review by srehaider — CHANGES_REQUESTED** [review]
Reviewed some files please make these changes.

File: src/pages/HomePage.js [hunk 8]
```diff
@@ -1,12 +1,20 @@
...
+}
   
   logoutUser = () => {
     window.localStorage.setItem('blogSiteUserLoggedIn', false);
-    this.props.changeLoginState(false);
-    this.props.navigator('/login');
+    logoutRequest(() => {
+      this.props.changeLoginState(false);
+      this.props.navigator('/login');
+    }).then();
```
Comment (srehaider) [inline src/pages/HomePage.js]: Did you verify it works?

Comment (musanaeem) [reply src/pages/HomePage.js]: Apologies, I thought i had made this change in this PR. I made it in the GET request PR. Changes made and uploaded.

**Review by srehaider — APPROVED** [review]

**Pull request merged**